Workiva/over_react
# FEDX-165 Upgrade to analyzer 5

## Motivation
The `analyzer` dependency in over_react and its analyzer plugin are outdated and need to be upgraded. We want to jump from analyzer 2 to analyzer 5.

## Changes
- Update CI's analyzer matrix, start running on 2.19

### Root over_react package
- Update to analyzer 5 and update related dependencies
- Address analyzer breakages, notably:
  - Removal of ClassOrMixinDeclaration
       - In most cases this was resolved by leveraging the existing ClassishDeclaration utility class that adapts ClassDeclaration, MixinDeclaration, and ClassTypeAlias (`class Foo = Bar with Baz`) into a single API
  - Change of name from Identifier to Token in many cases
  - Change from TypeName to NamedType
  - Removal of getMethod
- Update comment placement in generated code so that it formats consistently in dart_style versions resolved to in Dart 2.18 and 2.19
    - This made it easier for us to verify in CI that there are no generated changes that aren't checked in, and also the placement is an improvement over the old placement


### Analyzer plugin subpackage  
- Update to analyzer 5 and update related dependencies (notably including analyzer_plugin)
- Address analyzer breakages, notably:
  - Change of name from Identifier to Token in many cases
    - This affected NodeLocator/NodeLocator2 which were copied from the analyzer package, so those implementations were synced with their analyzer 5 versions 
  - Removal of ClassOrMixinDeclaration 
  - Elements for mixins are no longer ClassElement, but rather its supertype, InterfaceElement
  - References to types are no longer visited as Identifier, but rather non-Expression NamedType
    - This mainly affected exhaustive_deps, which needed to be updated to handle both of these cases
- Fix some built-in analysis warnings in test cases
- Address analyzer_plugin breakages. 

    Mainly, they moved away from having ServerPlugin subclasses instantiate their own `AnalysisDriver`s, in favor of the base class managing its own `AnalysisContextCollection` ([link to main changes](https://github.com/dart-lang/sdk/commit/46b73c67859162204cc9fe9c10d82e18aad84674)). This caused our analyzer setup both in the main plugin and in tests to become invalid, and needed to be refactored.
    
    Changes to address this:

    - Remove invalid `createAnalysisDriver` override, and instead use the `AnalysisContextCollection` provided by the base class
    - Wire up diagnostics to the new `analyzeFile` hook, which also seems to be called for non-"priority" files, allowing us to remove the workarounds we had in place to ensure diagnostics were run on all files within a package (see `_updatePriorityFiles` removal).
    - Refactor tests to use a slimmed down copy of over_react_codemod's `SharedAnalysisContext`. More details:
    
        Our tests were pretty tightly coupled to that old implementation, and involved creating our own analysis driver. This setup involved a mock Dart SDK, a memory-based resource provider that required copying code needed for resolved analysis (like over_react and transitive dependencies) and a custom package config.
          
        This setup was pretty complicated,  involved src imports, and also needed to be migrated in response to Dart SDK changes. As opposed to trying to update it, I opted to refactor these to use the approach we use in over_react_codemod's tests, which involves creating real `AnalysisContextCollection`s and sharing them across tests for performance. 


#### Release Notes
  

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
        - Verify in playground that:
            - Diagnostics, fixes, and assists work at a base level
            - Diagnostics, fixes, and assists work after when modifying files, to ensure the plugin isn't using stale analysis results. This is best tested by adding new code to an existing file.
        - Verify in a larger repo that diagnostics show for all files, and not just "priority files" (viewed by opening Analyzer Diagnostics from your IDE, selecting the "Contexts" tab, and scrolling down to the "Priority files" section)

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
